qiskit/transpiler/passes/utils/remove_final_measurements.py

```diff
@@ -92,17 +92,25 @@ def run(self, dag):
             # no idle wires to remove
             return dag
 
-        # determine unused cregs resulting from the removal
-        cregs_to_remove = set()
+        # determine bits of all registers where register is now idle
+        # as a result of the removal.
+        bits_to_remove = set()
+        bits_to_keep = set()
         for creg in dag.cregs.values():
             clbits = set(creg)
             if not clbits.isdisjoint(clbits_with_final_measures) and clbits.issubset(idle_wires):
-                cregs_to_remove.add(creg)
+                # register contains a newly idle bit, and all other bits are idle.
+                bits_to_remove |= clbits
+            else:
+                # register does not contain a newly idle bit, or contains other busy bits
+                # and thus should not be removed.
+                bits_to_keep |= clbits
 
-        # remove cregs from DAG
-        dag.remove_cregs(*cregs_to_remove)
+        # keep bits of registers we'll keep
+        bits_to_remove -= bits_to_keep
 
-        # remove any remaining unused individual clbits
-        dag.remove_clbits(*clbits_with_final_measures)
+        # also remove bits that aren't in a register at all
+        bits_to_remove |= clbits_with_final_measures - (bits_to_remove | bits_to_keep)
```

Comment:
I'll have to think about this a little bit, but is there ever a case where a bit in `clbits_with_final_measures` should not end up in `bits_to_remove`?

Reply:
A bit from `clbits_with_final_measures` should not end up in `bits_to_remove` if any of
- it is not idle after final measures are removed.
- it is in a register that is not idle after final measures are removed.

On line 89, we get rid of any bits that are not idle after final measures are removed. On this line (113), we get rid of any bits that are in a register.

Reply:
I agree with the code, but I think your line numbers are all a bit backwards there?  It looks to me like line 110 removes bits that are in registers which are not entirely idle, and line 113 removes bits that aren't in a register at all.

Thinking again, I think the `bits_to_remove` variable is unnecessary within the loop; the clbits you want to return here just need to satisfy "is idle" and "is not in a register that is not idle".  Taking every variable with its state after its first initialisation, that's just the set
```python
idle_clbits_with_final_measures = clbits_with_final_measures.intersection(self.idle_wires())
bits_to_remove = idle_clbits_with_final_measures - bits_to_keep
```
So I think there's no need to build up `bits_to_remove` within the `creg` loop - you can go straight to it in one go after creating `bits_to_keep` since you already need `idle_clbits_with_final_measures` to make `bits_to_keep`, and `bits_to_keep` might be more descriptive as `bits_in_busy_registers`.

By the way (off-topic and not important anyway), how come you sometimes use the method names (`a.intersection_update(b)` not `a &= b`) and sometimes the operators (`a |= b` not `a.update(b)` and `a -= b` not `a.difference_update(b)`) when you're always using sets?  The operators need set objects, and the methods can take general iterables, but in these cases everything's always a set.

Reply:
I think one subtlety here that isn't obvious is that there are actually bits we want to remove that are _not_ in `clbits_with_final_measures`. Specifically, if a register has at least one bit that has become idle after removing measurements, and all of its bits are now idle, we want to remove the entire register, and all of its bits. Some of those bits might not be in `clbits_with_final_measures`, i.e. they were idle before the original measure.

`bits_to_remove` allows us to collect those bits that are in a register that is now idle as a result of the final measurement removal including bits that aren't in `clbits_with_final_measures`.

I agree that some better naming might go a long way and will make some changes to that end.

> how come you sometimes use the method names

This was mostly just in line with my level of comfort for seeing those operations in the wild, and I'm happy to change it for consistency if you prefer. Personally, I'm comfortable seeing for example `|` and `-` for union and difference, but I would flinch a little bit seeing something like `<=` or `<` for subset and proper subset, at least as a brand new reader. With the method versions accepting iterables as well in some cases, and some operations (e.g. `isdisjoint`) not having an operator equivalent, I'm not sure exactly where we should draw the line.

Reply:
Ah yeah, I missed that extra case with `bits_to_remove`. That's definitely needed, then.

Reply:
Naming and temporary use improved in 96f1d7a.

Reply:
Yeah, I agree that the operators are way more mysterious.  Consistency either way is good, so I'm happy.